File: Level 2/방문 길이/solution.js
function solution(dirs) {
  const dirsMap = {
    U: [0, 1],
    D: [0, -1],
    R: [1, 0],
    L: [-1, 0],
  };

  let character = [0, 0];
  const routes = [];

  [...dirs].forEach((dir) => {
    const route = [character];
    const [dX, dY] = dirsMap[dir];
    const [nX, nY] = [character[0] + dX, character[1] + dY];

    if (nX >= -5 && nX <= 5 && nY >= -5 && nY <= 5) {
      character = [nX, nY];
      route.push(character);
      if (checkRoute(route, routes)) {
        routes.push(route);
      }
    }
  });

  return routes.length;
}

function checkRoute(value, routes) {
  for (const route of routes) {
    if (
      JSON.stringify(value) === JSON.stringify(route) ||
      JSON.stringify(value) === JSON.stringify([...route].reverse())
    ) {
      return false;
    }
  }

  return true;
}

File: Level 2/방문 길이/solution.py
def check_is_valid_move(x, y):
    return -5 <= x <= 5 and -5 <= y <= 5


def move(x, y, dir):
    if dir == "U":
        return [x, y + 1]
    elif dir == "L":
        return [x - 1, y]
    elif dir == "R":
        return [x + 1, y]
    elif dir == "D":
        return [x, y - 1]


def solution(dirs):
    location = [0, 0]
    routes = set()

    for dir in dirs:
        x, y = location
        nx, ny = move(x, y, dir)

        if check_is_valid_move(nx, ny):
            routes.add((x, y, nx, ny))
            routes.add((nx, ny, x, y))
            location = [nx, ny]

    return len(routes) / 2
